iXBRLViewerPlugin/viewer/src/js/factset.js

```diff
@@ -0,0 +1,98 @@
+// limitations under the License.
+
+export function FactSet(facts) {
+    this._facts = facts;
+}
+
+/* Returns the union of dimensions present on facts in the set */
+FactSet.prototype._allDimensions = function() {
+    var dims = {};
+    for (var i = 0; i < this._facts.length; i++) {
+        var dd = Object.keys(this._facts[i].dimensions());
+        for (var j = 0; j < dd.length; j++) {
+            dims[dd[j]] = true;
+        }
+    }
+    return Object.keys(dims);
+}
+
+/* Returns the "minimally unique" label for the specified fact in the set.
+ * 
+ * Minimally unique means that we include the value for just enough dimensions
+ * to generate labels that are unique within the set.
+ *
+ * In order to generate the best labels, we try concept and period first.
+ */
+FactSet.prototype.minimallyUniqueLabel = function(fact) {
+    if (!this._minimallyUniqueLabels) {
+        var allLabels = {};
+        var allAspects = ["c", "p"].concat(this._allDimensions());
+        /* Assemble a map of arrays of all aspect labels for all facts, in a
+         * consistent order */
+        for (var i = 0; i < this._facts.length; i++) {
+            var f = this._facts[i];
+            allLabels[f.id] = [];
+            for (var j = 0; j < allAspects.length; j++) {
+                var dd = f.aspects()[allAspects[j]];
+                allLabels[f.id].push(dd ? dd.valueLabel() : null);
+            }
+        }
+        /* Iterate each aspect label and compare that label across all facts in
+         * the set */
+        var uniqueLabels = {};
+        for (var j = 0; j < allAspects.length; j++) {
+            var labelMap = {};
+            for (var i = 0; i < this._facts.length; i++) {
+                labelMap[allLabels[this._facts[i].id][j]] = true;
+            }
+
+            var uniqueLabelsByLabel = {};
+            if (Object.keys(labelMap).length > 1) {
+                /* We have at least two different labels, so include this
+                 * aspect in the label for all facts in the set */
+                for (var i = 0; i < this._facts.length; i++) {
+                    var fid = this._facts[i].id;
+                    var l = allLabels[fid][j];
+                    var ul = uniqueLabels[fid] || [];
+                    if (l !== null) {
+                        ul.push(l);
+                    }
+                    if (ul.length > 0) {
+                        uniqueLabels[fid] = ul;
+                        uniqueLabelsByLabel[ul.join(", ")] = true;
+                    }
+                } 
+                /* We have as many different labels as facts - we're done */
+                if (Object.keys(uniqueLabelsByLabel).length == this._facts.length) {
+                    break;
+                }
+            }
+        }
+
+        /* If any label is empty, add the concept label onto the start of all
+         * of them */
+        if (Object.keys(uniqueLabels).length < this._facts.length) {
```

Comment:
I think my misunderstanding [here](https://github.com/Workiva/ixbrl-viewer/pull/34#discussion_r288124188) is overloading of "label" in this function to mean both "single aspect value for a single fact" and "concatenation of all non-redundant aspect values for a single fact".  I was misreading this comment as "if any aspect value for any fact", but you mean "if any fact has either empty or redundant values for all aspects", right?  And since concept is a required aspect, the only way to get here is if all facts have the same concept but no value (or no non-redundant values) for other aspects, so we can just add the concept to the front.